### PR TITLE
Basic Integration

### PR DESCRIPTION
**Pull request by llravelo**
# Description

Asana Basic Integration

**Notes**:
- An NPM package for Asana [exists](https://www.npmjs.com/package/asana), however there some issues with it such as:
  1. Version divergence between published NPM package and github tag.
  2. Last publish is over a year ago, with git updates mostly consisting of OpenAPI syncs.
  3. The package itself has deprecated dependencies ([request](https://www.npmjs.com/package/request), [har-validator](https://www.npmjs.com/package/har-validator))

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [x] This change requires a documentation update

### General Development Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [x] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [x] My changes properly paginate the target service provider's API
- [x] My changes properly handle rate limiting of the target service provider's
      API
- [x] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [x] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [x] I have updated the `CHANGELOG.md` file to describe my changes
- [x] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: src/steps/user/converter.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+export function createUserEntity(data: User): Entity {
+  return createIntegrationEntity({
+    entityData: {
+      source: data,
+      assign: {
+        _key: getUserKey(data.gid),
+        _type: Entities.USER._type,
+        _class: Entities.USER._class,
+        username: data.email,
+        active: true,
```
Comment (llravelo) [inline src/steps/user/converter.ts]: @eXtremeX There was no `active` property in the most recent [endpoint](https://developers.asana.com/docs/get-multiple-users) so I just assumed that all existing users were active. Same case for username, I just used existing email property.

File: docs/jupiterone.md [hunk 2]
```diff
@@ -96,21 +96,30 @@ https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
...
-| Source Entity `_type` | Relationship `_class` | Target Entity `_type` |
-| --------------------- | --------------------- | --------------------- |
-| `acme_account`        | **HAS**               | `acme_group`          |
-| `acme_account`        | **HAS**               | `acme_user`           |
-| `acme_group`          | **HAS**               | `acme_user`           |
+| Source Entity `_type`      | Relationship `_class` | Target Entity `_type`      |
+| -------------------------- | --------------------- | -------------------------- |
+| `asana_account`            | **HAS**               | `asana_workspace`          |
+| `asana_project`            | **HAS**               | `asana_project_membership` |
+| `asana_project_membership` | **HAS**               | `asana_user`               |
```
Comment (eXtremeX) [inline docs/jupiterone.md]: I'd change this slightly to be:

`asana_user` -ASSIGNED-> `asana_project_membership`.

`asana_project_membership` acts as permissions here and generally users have assigned certain permissions.

File: docs/jupiterone.md [hunk 3]
```diff
@@ -96,21 +96,30 @@ https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
...
+| `asana_team`               | **HAS**               | `asana_project`            |
```
Comment (eXtremeX) [inline docs/jupiterone.md]: I'm also thinking if:

`asana_team` -ASSIGNED-> `asana_project`

would be appropriate here. In plain English it does seem as if some team is assigned to particular project and isn't necessarily parent of it.

File: docs/jupiterone.md [hunk 4]
```diff
@@ -96,21 +96,30 @@ https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
...
-| User      | `acme_user`    | `User`          |
-| UserGroup | `acme_group`   | `UserGroup`     |
+| Resources          | Entity `_type`             | Entity `_class` |
+| ------------------ | -------------------------- | --------------- |
+| Account            | `asana_account`            | `Account`       |
+| Project            | `asana_project`            | `Project`       |
+| Project Membership | `asana_project_membership` | `AccessRole`    |
+| Team               | `asana_team`               | `Team`          |
+| User               | `asana_user`               | `User`          |
+| Workspace          | `asana_workspace`          | `Organization`  |
```
Comment (eXtremeX) [inline docs/jupiterone.md:98]: @llravelo Do you know if there can be more than one workspace for the single account/auth? If no, `asana_workspace` feels like perfect root node (even right now if you take a look at graph it feels more centralized than `asana_account`) but it really depends on this.

image

**Review by eXtremeX — CHANGES_REQUESTED** [review]
Hey @llravelo looks extremely good! I haven't done super detailed PR review just yet, but I do want to make you aware of a few ideas/questions I've asked so I'm submitting this as is and will continue on reviewing it.

File: src/client.ts [hunk 5]
```diff
@@ -1,121 +1,197 @@
...
+  public async iterateWorkspaces(iteratee: ResourceIteratee<Workspace>) {
+    let body: Paginated<Workspace[]>;
+    let path = `/workspaces?limit=${this.paginateEntitiesPerPage}&opt_fields=gid,name,resource_type`;
+    let endpoint: string;
+
+    do {
+      endpoint = this.withBaseUri(path);
+      body = await this.request<Paginated<Workspace[]>>(endpoint);
+      for (const workspace of body.data) await iteratee(workspace);
+      if (body.next_page !== null) path = body.next_page.path;
```
Comment (eXtremeX) [inline src/client.ts]: We would like to see these changed to the following (throughout the code):

```suggestion
      for (const workspace of body.data) {
         await iteratee(workspace);
      }
      if (body.next_page !== null) {
         path = body.next_page.path;
      }
```

It's just all other integrations follow that style so we'd like to stick with that.

File: docs/jupiterone.md (hunk 4 above)
Comment (llravelo) [reply docs/jupiterone.md:98]: @eXtremeX Yes, it is possible to have more than one workspace. For example, I just created a new workspace with sample members. This is how it looks like now:

image

File: docs/jupiterone.md (hunk 2 above)
Comment (llravelo) [reply docs/jupiterone.md]: Ok, will do.

File: src/client.ts (hunk 5 above)
Comment (llravelo) [reply src/client.ts]: Sure!

File: docs/jupiterone.md (hunk 2 above)
Comment (llravelo) [reply docs/jupiterone.md]: The `/projects/{projectId}/project_memberships` payload roughly looks like this.
```
{
    "gid": "{gid_number}",
    "project": {
        "gid": "{gid_number}",
        "resource_type": "project"
    },
    "user": {
        "gid": "{gid_number}",
        "resource_type": "user"
    },
    "write_access": "{permission}",
    "resource_type": "project_membership"
},
```
I think it would also make sense to add `PROJECT_MEMBERSHIP ALLOWS PROJECT` in this case, unless it's unecessary.

File: docs/jupiterone.md (hunk 3 above)
Comment (llravelo) [reply docs/jupiterone.md]: I agree also, it makes more sense that way.

**Review by neurovelho — APPROVED** [review]
